contracts/src/utils/structs/checkpoints.rs

```diff
@@ -0,0 +1,463 @@
+//! define a variable type [`Trace160`] in your contract, and store a
+//! new checkpoint for the current transaction block using the
+//! [`Trace160::push`] function.
+use alloy_primitives::{uint, Uint, U256, U32};
+use alloy_sol_types::sol;
+use stylus_proc::{sol_storage, SolidityError};
+use stylus_sdk::storage::{StorageGuard, StorageGuardMut};
+
+use crate::utils::math::alloy::Math;
+
+// TODO: add generics for other pairs (uint32, uint224) and (uint48, uint208).
+// Logic should be the same.
+type U96 = Uint<96, 2>;
+type U160 = Uint<160, 3>;
+
+sol! {
+    /// A value was attempted to be inserted on a past checkpoint.
+    #[derive(Debug)]
+    error CheckpointUnorderedInsertion();
+}
+
+/// An error that occurred while calling [`Trace160`] checkpoint contract.
+#[derive(SolidityError, Debug)]
+pub enum Error {
+    /// A value was attempted to be inserted on a past checkpoint.
+    CheckpointUnorderedInsertion(CheckpointUnorderedInsertion),
+}
+
+sol_storage! {
+    /// State of checkpoint library contract.
+    #[cfg_attr(all(test, feature = "std"), derive(motsu::DefaultStorageLayout))]
+    pub struct Trace160 {
+        /// Stores checkpoints in a dynamic array sorted by key.
+        Checkpoint160[] _checkpoints;
+    }
+
+    /// State of a single checkpoint.
+    pub struct Checkpoint160 {
+        /// Key of checkpoint. Used as a sorting key.
+        uint96 _key;
+        /// Value corresponding to the key.
+        uint160 _value;
+    }
+}
+
+impl Trace160 {
+    /// Pushes a (`key`, `value`) pair into a Trace160 so that it is
+    /// stored as the checkpoint.
+    ///
+    /// Returns previous value and new value.
+    ///
+    /// IMPORTANT: Never accept `key` as a user input, since an arbitrary
+    /// `U96::MAX` key set will disable the library.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the checkpoint's state.
+    /// * `key` - Last checkpoint key to insert.
+    /// * `value` - Checkpoint value corresponding to insertion `key`.
+    ///
+    /// # Errors
+    ///
+    /// If the `key` is lower than previously pushed checkpoint's key error
+    /// [`Error::CheckpointUnorderedInsertion`] is returned (necessary to
+    /// maintain sorted order).
+    pub fn push(
+        &mut self,
+        key: U96,
+        value: U160,
+    ) -> Result<(U160, U160), Error> {
+        self._insert(key, value)
+    }
+
+    /// Returns the value in the first (oldest) checkpoint with key
+    /// greater or equal than the search key, or zero if there is none.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    /// * `key` - Checkpoint's key to lookup.
+    pub fn lower_lookup(&self, key: U96) -> U160 {
+        let len = self.length();
+        let pos = self._lower_binary_lookup(key, U256::ZERO, len);
+        if pos == len {
+            U160::ZERO
+        } else {
+            self._access(pos)._value.get()
+        }
```

Comment:
Well, all of this could be replaced with [`core::slice::binary_search_by`](https://doc.rust-lang.org/beta/core/primitive.slice.html#method.binary_search_by), right?

I think we can simplify this implementation, but then again, we need to have ways to measure gas. Do you mind creating an issue for exploring this?

Reply:
I'm not sure that stylus sdk StorageVec is slice or has similar api. It seems for me we should have a binary search written by ourselves...

Reply:
Ah, you're right, I forget we're not dealing with a `Vec` here.